crates/pathfinder/src/storage/schema/revision_0010.rs

```diff
@@ -298,6 +293,254 @@ mod tests {
+                        .collect()
+                }
+
+                fn event_key_to_base64_string(key: &EventKey) -> String {
+                    base64::encode(key.0.as_be_bytes())
+                }
+
+                pub fn event_keys_to_base64_strings(keys: &[EventKey]) -> String {
+                    // TODO: we really should be using Iterator::intersperse() here once it's stabilized.
+                    let keys: Vec<String> =
+                        keys.iter().map(Self::event_key_to_base64_string).collect();
+                    keys.join(" ")
+                }
+
+                pub fn insert_events(
+                    connection: &Connection,
+                    block_number: StarknetBlockNumber,
+                    transaction: &transaction::Transaction,
+                    events: &[transaction::Event],
+                ) -> anyhow::Result<()> {
+                    use anyhow::Context;
+
+                    if transaction.contract_address.is_none() && !events.is_empty() {
+                        anyhow::bail!("Declare transactions cannot emit events");
+                    }
+
+                    for (idx, event) in events.iter().enumerate() {
+                        connection
+                        .execute(
+                            r"INSERT INTO starknet_events ( block_number,  idx,  transaction_hash,  from_address,  keys,  data)
+                                                   VALUES (:block_number, :idx, :transaction_hash, :from_address, :keys, :data)",
+                            rusqlite::named_params![
+                                ":block_number": block_number.0,
+                                ":idx": idx,
+                                ":transaction_hash": &transaction.transaction_hash.0.as_be_bytes()[..],
+                                ":from_address": &event.from_address.0.as_be_bytes()[..],
+                                ":keys": Self::event_keys_to_base64_strings(&event.keys),
+                                ":data": Self::event_data_to_bytes(&event.data),
+                            ],
+                        )
+                        .context("Insert events into events table")?;
+                    }
+                    Ok(())
+                }
+
+                pub(crate) const PAGE_SIZE_LIMIT: usize = 1024;
+
+                pub fn get_events(
+                    connection: &Connection,
+                    filter: &StarknetEventFilter,
+                ) -> anyhow::Result<PageOfEvents> {
+                    use anyhow::Context;
+
+                    let mut base_query =
+            r#"SELECT
+                  block_number,
+                  starknet_blocks.hash as block_hash,
+                  transaction_hash,
+                  from_address,
+                  data,
+                  starknet_events.keys as keys
+               FROM starknet_events
+               INNER JOIN starknet_blocks ON starknet_blocks.number = starknet_events.block_number "#
```

Comment:
Wasn't the join condition supposed to be in parentheses?

Reply:
Oh wait this is some backed up version. Why is this all backed up?

Reply:
Everytime we change the production code; the older migrations break / become suspect. And often this breakage is silent.